instrumentation/opentelemetry-instrumentation-boto3sqs/src/opentelemetry/instrumentation/boto3sqs/__init__.py

```diff
@@ -375,52 +379,56 @@ def _wrap_client_creation(self) -> None:
 
         def client_wrapper(wrapped, instance, args, kwargs):
             retval = wrapped(*args, **kwargs)
-            if not self._did_decorate:
-                self._decorate_sqs()
+            self._decorate_sqs(type(retval))
             return retval
 
         wrap_function_wrapper(boto3, "client", client_wrapper)
 
-    def _decorate_sqs(self) -> None:
+    def _decorate_sqs(self, sqs_class: type) -> None:
         """
         Since botocore creates classes on the fly using schemas, we try to find the class that inherits from the base
         class and is SQS to wrap.
         """
         # We define SQS client as the only client that implements send_message_batch
-        sqs_class = [
-            cls
-            for cls in botocore.client.BaseClient.__subclasses__()
-            if hasattr(cls, "send_message_batch")
-        ]
-        if sqs_class:
-            self._sqs_class = sqs_class[0]
-            self._did_decorate = True
-            self._wrap_send_message()
-            self._wrap_send_message_batch()
-            self._wrap_receive_message()
-            self._wrap_delete_message()
-            self._wrap_delete_message_batch()
-
-    def _un_decorate_sqs(self) -> None:
-        if self._did_decorate:
-            unwrap(self._sqs_class, "send_message")
-            unwrap(self._sqs_class, "send_message_batch")
-            unwrap(self._sqs_class, "receive_message")
-            unwrap(self._sqs_class, "delete_message")
-            unwrap(self._sqs_class, "delete_message_batch")
-            self._did_decorate = False
+        if not hasattr(sqs_class, "send_message_batch"):
+            return
+
+        if getattr(sqs_class, _IS_SQS_INSTRUMENTED_ATTRIBUTE, False):
+            return
```

Comment:
I think this caching mechanism is not needed. There cannot be two `botocore.client.BaseClient` anyway, so if you instrumented once, you will not be able to instrument again (because of the instrumentor's base class caching mechanism)

Reply:
While it is true that there is only one `botocore.client.BaseClient`, there can be multiples of any class that is generated by botocore, like `botocore.client.SQS`.  When calling `boto3.client(...)`, `botocore` will internally [generate](https://github.com/boto/botocore/blob/1.27.56/botocore/client.py#L174) the concrete service class.  It will do so every time even if the class for an AWS service was already created before. This means, when you call `boto3.client("sqs")` two or more times you will get instances that are type `botocore.client.SQS` but their actual `botocore.client.SQS` type is not the same.
The `_IS_SQS_INSTRUMENTED_ATTRIBUTE` attribute is an indicator for the instrumentor to know which class was already instrumented. It is only set onto the concrete/generated class returned by the `boto3.client(...)` call. The `botocore.client.BaseClient` is never modified and does not have this attribute at any time.

Your comment made me think that `_IS_SQS_INSTRUMENTED_ATTRIBUTE` should be reset/unset in the `_un_decorate_sqs` function as otherwise the concrete class will not instumented again in a instrument -> uninstrument -> instrument call sequence.

Reply:
You're right. I now understand your solution :)